### ISSUE TITLE
Add an MCMCResult object

### ISSUE THREAD

**Issue by chonlei** [pints-team/pints#129]
Maybe work with the save_to_file branch?

**Comment by MichaelClerx** [comment]
Agreed we need to make MCMC methods return an object, but I think it shouldn't be related to storing results on disk (or storing partial results on disk so we can restart again), those are separate issues!

**Comment by MichaelClerx** [comment]
See also #137

**Comment by MichaelClerx** [comment]
> I think it shouldn't be related to storing results on disk (or storing partial results on disk so we can restart again), those are separate issues!

Then again, we might need to let users choose between a "heavy" and a "lightweight" results object to avoid memory issues. This would have to be worked out in #132

**Comment by MichaelClerx** [comment]
This is causing the tests to fail!

**Comment by ben18785** [comment]
Yes, it's the dependence on 'tabulate'. Can we add this as a dependency
please? I wrote about this in the pull request.

**Comment by MichaelClerx** [comment]
Ok, fixed the test. Not sure about adding tabulate as a dependency! Will have a look at this this week!

**Comment by ben18785** [comment]
Ok, am definitely aware that the print output which tabulate produces could
be handled without an extra dependency. I just think that there's a
trade-off between dependencies-needed and time taken to develop a
functionality. But happy if someone else wants to mimic the tabulate result
in pure Python.

**Comment by MichaelClerx** [comment]
Ok. Will close this again as it has already been merged. Will add separate tickets for remaining issues.

Would really have like to see this in a branch until

- all tests passed
- new tests were written specifically for this
- documentation had been added
- all methods had been adapted to return these objects
- the examples had been updated

**Comment by MichaelClerx** [comment]
Okay, re-opening this again for some discussion :-)

What exactly is the purpose of the returned object? The current version just wraps around a numpy array and does some diagnostics, is that all we need? Because in that case I'd prefer just to return a numpy array.

What other functionality / information do we want it to provide?

Is there some kind of type-checking involved? Do we envision passing this object on to other methods?

**Comment by ben18785** [comment]
Sorry, but again this is something I feel passionately about. Some of the
things we are suggesting doing are making it dangerously easy for a user to
mess up doing MCMC.

I can't stress enough how important having these diagnostics there, by
default, is. If you don't know R-hat (i.e. whether your algorithm has
converged) then you can't decide whether or not to take the numpy array, or
(as is usually the case) to run the chains for longer. These convergence
issues are likely to be even more important for ODE models than in standard
statistical models.

There's a reason literally ALL other packages that exist and have ever
existed return these objects. It's because MCMC is not foolproof. It has
many pitfalls that people will walk into unknowingly.

This can't be addressed by simply including a separate Rhat calculation in
all of our examples. People being lazy will ignore them.

**Comment by MichaelClerx** [comment]
People being lazy can ignore the result object as well, or run only 1 iteration?

But that doesn't answer my question. In your view, is the purpose of the MCMCResult object to perform diagnostics on one or multiple chains?

**Comment by ben18785** [comment]
Yes, but we can nudge them in the right direction by making things the
default option.

It's main function is to act as a barrier to people's ignorance. To stop
them thinking "I've run a single chain for 1000 iterations, it must have
converged by now." I can't stress how imperfect MCMC is. There are
literally books about how bad the method can be if you don't know what to
look for. Because of this, having a return object that contains all this
information by default is completely standard. If we didn't do it, people
would think we are lagging behind. I'm sorry, but I won't budge on this.

**Comment by MichaelClerx** [comment]
That's fine! I just need to know _why_ we want these things so I can wrap my head around it and try to design something that meets all our current (and some of our future) needs.
If we want to do these things in a performance oriented way, or simply in a flexible way, it's entirely possible that things will be written to disk before being analysed with any kind of statistic, or that a user will want thinning during a run to save memory, or that we want to use these diagnostics during a run to decide if we can stop or not.

None of this precludes having some default option that runs with default parameters and returns a result with some safety options. But runs with e.g. ephys models can easily take days, when running those, we should assume our users are more aware of possible dangers and will want more control

**Comment by MichaelClerx** [comment]
I also don't understand why wrapping the results in an object that also does statistics is considered safe. What's to stop users from assuming these diagnostics are just some internal detail and ignoring them?

**Comment by MichaelClerx** [comment]
See #986
